src/trust/chain.ts

```diff
@@ -114,38 +85,51 @@ export async function validateTrustChain(
  *
  * @param chain The original chain
  * @param appFetch (optional) fetch api implementation
- * @returns A list of signed token that reprensent the trust chain, in the same order of the provided chain
- * @throws When an element of the chain fails to parse
+ * @returns A list of signed token that represent the trust chain, in the same order of the provided chain
+ * @throws IoWalletError When an element of the chain fails to parse
  */
-export function renewTrustChain(
+export async function renewTrustChain(
   chain: string[],
   appFetch: GlobalFetch["fetch"] = fetch
-) {
+): Promise<string[]> {
   return Promise.all(
-    chain
-      // Decode each item to determine its shape
-      .map(decode)
-      .map(
-        (e) =>
-          [
-            EntityStatement.safeParse(e),
-            EntityConfiguration.safeParse(e),
-          ] as const
-      )
-      // fetch the element according to its shape
-      .map(([es, ec], i) =>
-        ec.success
-          ? getSignedEntityConfiguration(ec.data.payload.iss, { appFetch })
-          : es.success
-          ? getSignedEntityStatement(es.data.payload.iss, es.data.payload.sub, {
-              appFetch,
-            })
-          : // if the element fail to parse in both EntityStatement and EntityConfiguration, raise an error
-            Promise.reject(
-              new IoWalletError(
-                `Cannot renew trust chain because the element #${i} failed to be parsed.`
-              )
-            )
-      )
+    chain.map(async (token, index) => {
+      const decoded = decode(token);
+
+      const entityStatementResult = EntityStatement.safeParse(decoded);
+      const entityConfigurationResult = EntityConfiguration.safeParse(decoded);
+
+      if (entityConfigurationResult.success) {
+        return getSignedEntityConfiguration(
+          entityConfigurationResult.data.payload.iss,
+          { appFetch }
+        );
+      } else if (entityStatementResult.success) {
```

Comment:
Since we are returning early we can remove the `else if` and just use `if`, what do you think?